### PR TITLE
added veDeviceMesh

### PR DESCRIPTION
**Pull request by MackZackA**
This PR introduces veDeviceMesh, the device mesh API that integrates handling of submeshes and process groups in performing training with DDP, TP/SP, distributed optimizer and checkpointing. It also updates fixes and patches related to veDeviceMesh API to the repository since last PR.

### CODE REVIEW

File: python/vescale/dtensor/sharding_spec.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+################################################################################
```
Comment (MingjiHan99) [inline python/vescale/dtensor/sharding_spec.py:1]: I thinks this class have been removed. https://github.com/volcengine/veScale/pull/23

File: python/example/nanogpt_4D_finetune/finetune_4D.py [hunk 2]
```diff
@@ -30,11 +30,11 @@
```
Comment (leonardo0lyj) [inline python/example/nanogpt_4D_finetune/finetune_4D.py:30]: `python` folder should be gone by a picked MR?

File: test/parallel/ddp_optim/test_grad_sync.py [hunk 3]
```diff
@@ -192,7 +165,6 @@ def test_ddp(self, overlap_grad_reduce: bool, use_distributed_optimizer: bool):
             m,
             tp_submesh,
             {"parameter": param_sharding_plan, "forward": fwd_sharding_plan},
-            grad_sync={torch.nn.LayerNorm: ["weight", "bias"]},
         )
 
         ddp_m = DDP(
```
Comment (leonardo0lyj) [inline test/parallel/ddp_optim/test_grad_sync.py:170]: where is `test/*/devicemesh_api`?

File: python/example/nanogpt_4D_finetune/finetune_4D.py (hunk 2 above)
Comment (MackZackA) [reply python/example/nanogpt_4D_finetune/finetune_4D.py:30]: yes. it is done in the latest commit after incorporating randomness.

File: test/parallel/ddp_optim/test_grad_sync.py (hunk 3 above)
Comment (MackZackA) [reply test/parallel/ddp_optim/test_grad_sync.py:170]: incorporated in the second push after randomness.

File: third-party/torchdistx/.clang-tidy
Comment (leonardo0lyj) [inline third-party/torchdistx/.clang-tidy:1]: `third-party/*` should never show up，`torchdistx` is already built in Dockerfile

File: vescale/model/attention/dmodule_parallel_attention.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/model/attention/dmodule_parallel_attention.py:1]: remove `model/attention`; megatron code is not maintained in vescale

File: vescale/model/base_gpt/attention.py [hunk 5]
```diff
@@ -0,0 +1,531 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/model/base_gpt/attention.py:1]: remove `model/base_gpt`; megatron code is not maintained in vescale

File: vescale/model/random.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/model/random.py:1]: remove `model/random.py`; megatron code is not maintained in vescale

File: vescale/model/utils.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/model/utils.py:1]: remove `model/util.py`; megatron code is not maintained in vescale

File: vescale/optim/zero_redundant_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,1554 @@
+################################################################################
```
Comment (leonardo0lyj) [inline vescale/optim/zero_redundant_optimizer.py:1]: remove `vescale/optim/zero_redundant_optimizer`

File: test/oss/test_copyright.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
+################################################################################
```
Comment (leonardo0lyj) [inline test/oss/test_copyright.py:1]: remove this `oss` folder

**Review by Vremold — APPROVED** [review]

**Review by leonardo0lyj — APPROVED** [review]

**Pull request merged**